src/mpid/ch4/generic/am/mpidig_am_req_cache.h

```diff
@@ -0,0 +1,140 @@
+    MPIDIG_req_map_value_t *entry_list = NULL;
+    MPIDIG_req_map_value_t *list_head = NULL;
+    void *ret = NULL;
+
+    ret = MPIDIU_map_lookup(req_map, (uint64_t) sreq);
+    if (ret != MPIDIU_MAP_NOT_FOUND) {
+        entry_list = (MPIDIG_req_map_value_t *) ret;
+        if (entry_list->next == NULL) {
+            MPIDIU_map_erase(req_map, (uint64_t) sreq);
+            MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                            (MPL_DBG_FDEST, "req_map remove: entry_list=%p, entry=%p", entry_list,
+                             entry));
+        } else {
+            list_head = entry_list;
+            DL_FOREACH_SAFE(list_head, entry, tmp) {
+                if (entry->src_rank == src_rank) {
+                    MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                                    (MPL_DBG_FDEST, "req_map remove: entry_list=%p, entry=%p",
+                                     list_head, entry));
+                    DL_DELETE(list_head, entry);
+                    MPL_free(entry);
+                }
+            }
+            if (list_head != entry_list) {
+                MPIDIU_map_erase(req_map, (uint64_t) sreq);
+                MPIDIU_map_set(req_map, (uint64_t) sreq, (void *) list_head, MPL_MEM_OTHER);
+                MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                                (MPL_DBG_FDEST, "req_map remove: update entry_list=%p", list_head));
+            }
+        }
+    }
+}
+
+static inline void MPIDIG_req_cache_add(int handler_id, const void *am_hdr, MPIR_Request * rreq)
+{
+    const MPIDIG_hdr_t *hdr = (MPIDIG_hdr_t *) am_hdr;
+    if (!rreq || handler_id == MPIDIG_SEND || handler_id == MPIDIG_SEND_PIPELINE_SEG) {
+        return;
+    }
+    MPIDIG_req_map_add(MPIDIG_global.req_map, hdr->sreq_ptr, hdr->src_rank, rreq);
+    MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                    (MPL_DBG_FDEST, "req cache add: sreq=%p, src_rank=%d, rreq=%p handle=0x%x",
+                     hdr->sreq_ptr, hdr->src_rank, rreq, rreq->handle));
+}
+
+static inline void MPIDIG_req_cache_remove(int handler_id, const void *am_hdr)
+{
+    const MPIDIG_hdr_t *hdr = (MPIDIG_hdr_t *) am_hdr;
+    if (handler_id == MPIDIG_SEND || handler_id == MPIDIG_SEND_PIPELINE_SEG) {
+        return;
+    }
+    MPIDIG_req_map_remove(MPIDIG_global.req_map, hdr->sreq_ptr, hdr->src_rank);
+    MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                    (MPL_DBG_FDEST, "req cache remove: sreq=%p, src_rank=%d", hdr->sreq_ptr,
+                     hdr->src_rank));
+}
+
+static inline void MPIDIG_req_cache_find(int handler_id, const void *am_hdr, MPIR_Request ** rreq)
+{
+    const MPIDIG_hdr_t *hdr = (MPIDIG_hdr_t *) am_hdr;
+    MPIR_Request *request = NULL;
+    if (handler_id == MPIDIG_SEND || handler_id == MPIDIG_SEND_PIPELINE_SEG) {
+        *rreq = request;
+        return;
+    }
+    request = MPIDIG_req_map_lookup(MPIDIG_global.req_map, hdr->sreq_ptr, hdr->src_rank);
```

Comment:
I'm confused by the use of `sreq_ptr` here. If this is used after an RTS/CTS exchange, can't the `rreq_ptr` be sent directly by the sender process?

Reply:
For SEND operation, this is not needed as you can see from the `MPIDIG_req_cache_find()` which bypasses if the handler id is SEND or SEND_SEG.

For RMA this is needed as RMA does not send an CTS for incoming pipelined data, therefore we need this req cache to know if this incoming RMA message belongs to a RMA recv that is started.

Reply:
Also, I need to change the rma am_msg_t to include this common `MPIDIG_hdr_t` into it.

Reply:
Instead of differentiating the handling by handler_id, why not always have a `rreq_ptr` in the am_hdr, and set to 0 if sender doesn't have it. That way, the target callback can simply check `am_hdr` before decide whether to check cache. While effectively the same, I expect this way the code is simpler and more uniform.

Reply:
That would not work for RMA operations. For example, a pipelined MPI_PUT the rreq_ptr is only known when the first segment arrives on the target and MPI_PUT_TARGET_MSG_CB called. But unlike pipelined large message SEND which the sender always wait for CTS before sending the 2nd and rest segment, a pipelined RMA does not need to wait for the CTS, therefore the sender has no way to find out the rreq_ptr.

Reply:
> I'm confused by the use of `sreq_ptr` here. If this is used after an RTS/CTS exchange, can't the `rreq_ptr` be sent directly by the sender process?

Ah, I see. I had it backwards.

Reply:
> 
> 
> That would not work for RMA operations. For example, a pipelined MPI_PUT the rreq_ptr is only known when the first segment arrives on the target and MPI_PUT_TARGET_MSG_CB called. But unlike pipelined large message SEND which the sender always wait for CTS before sending the 2nd and rest segment, a pipelined RMA does not need to wait for the CTS, therefore the sender has no way to find out the rreq_ptr.

When the sender doesn't know `rreq_ptr` then simply set it to 0. I don't see why you have to differentiate so much between PT2PT and RMA. Both can either wait for CTS or don't wait, right? In my mind, they are just orthogonal.

Reply:
> > That would not work for RMA operations. For example, a pipelined MPI_PUT the rreq_ptr is only known when the first segment arrives on the target and MPI_PUT_TARGET_MSG_CB called. But unlike pipelined large message SEND which the sender always wait for CTS before sending the 2nd and rest segment, a pipelined RMA does not need to wait for the CTS, therefore the sender has no way to find out the rreq_ptr.
> 
> When the sender doesn't know `rreq_ptr` then simply set it to 0. I don't see why you have to differentiate so much between PT2PT and RMA. Both can either wait for CTS or don't wait, right? In my mind, they are just orthogonal.

PT2PT needs to wait for CTS. The sender would only know that the SEND is matched on the receiver side and user buffer is available for receiving all the data segments.

In RMA it does not need to wait for CTS. The origin already knows the user buffer on the target side, so it always can send into that buffer. While RMA can be done in the way of waiting for CTS, but this would just add a roundtrip overhead.

Reply:
You are hard coding it so PT2PT doesn't do eager pipeline and RMA does, thus tying two orthogonal concepts together.

Reply:
> You are hard coding it so PT2PT doesn't do eager pipeline and RMA does, thus tying two orthogonal concepts together.

The problem is for RMA to wait CTS is an unnecessary round trip. For PT2PT not waiting CTS, it would require the receiver to allocate large unexpected buffer. These two scenarios are undesirable. This is why we are doing eager(rts+data)-cts-pipeline_sendall for PT2PT, and pipeline_sendall for RMA.

For your reference, this is the same as the `main` branch. In main branch, these are the larger than eager threshold message transmission models today:

POSIX:
PT2PT (SEND): eager(rts-nodata)-cts-pipeline_sendall [into user buffer]
PT2PT (SSEND): pipeline_sendall [unmatched: receiver side create large unexpected buffer/ matched: into user buffer]
RMA: pipeline_sendall (into user buffer)

OFI:
PT2PT (SEND): eager(rts-nodata)-cts-rdma_read_sendall [contig:into user buffer/ unpack: allocate large unpack buffer, then unpack to user buffer]
PT2PT (SSEND): rdma_read_sendall [unmatched: receiver side create large unexpected buffer/ matched+contig: into user buffer / matched+unpack: into large unpack buffer, then unpack to user buffer]
RMA: pipeline_sendall [contig:into user buffer/ unpack: allocate large unpack buffer, then unpack to user buffer]

Reply:
Whether to do "EAGER" or "RNDZ" is an orthogonal decision from Whether to do "PIPELINE" or "SINGLE_PAYLOAD". We certainly can choose to always do `RNDZ-PIPELINE` and `EAGER-SINGLE`, but tying them in code just make it more complex.

> Also, I need to change the rma am_msg_t to include this common MPIDIG_hdr_t into it.

I would recommend not to nest the header structure, just duplicate the fields if you need them. If each handler_id associate with its independent header, then we can easily add or remove fields without worry of side effects (affecting other unrelated protocols).

Reply:
> Whether to do "EAGER" or "RNDZ" is an orthogonal decision from Whether to do "PIPELINE" or "SINGLE_PAYLOAD". We certainly can choose to always do `RNDZ-PIPELINE` and `EAGER-SINGLE`, but tying them in code just make it more complex.

The models I mentioned above is only for large message. And to be clear, what the main branch use today is not EAGER vs RNDV, it is SENDALL vs RNDV.

> 
> > Also, I need to change the rma am_msg_t to include this common MPIDIG_hdr_t into it.
> 
> I would recommend not to nest the header structure, just duplicate the fields if you need them. If each handler_id associate with its independent header, then we can easily add or remove fields without worry of side effects (affecting other unrelated protocols).

Yes, I am only adding `MPIDIG_hdr_t` into am msg_t structure of the RMA operations that need to transmit data. That means only. The involved ones are:
1. put/put_dt/put_dat
2. get/get_ack
3. acc/acc_dt/acc_dat
4. get_acc/get_acc_dt/get_acc_dat/get_acc_ack
5. cswap/cswap_ack

Reply:
> Yes, I am only adding MPIDIG_hdr_t into am msg_t structure of the RMA operations 

There are 4 fields in `MPIDIG_hdr_t`: `src_rank`, `tag`, `context_id`, `error_bits`. You only need `src_rank` for the RMA operations, right? Bring `MPIDIG_hdr_t` entirely to `RMA` , I would think to be really confusing.